graphql_compiler/compiler/ir_lowering_match.py

```diff
@@ -578,6 +593,194 @@ def prune_output_blocks_in_compound_match_query(compound_match_query):
+        elif expression.operator == u'&&':
+            if expression.left == TrueLiteral and expression.right == TrueLiteral:
+                return TrueLiteral
+            if expression.left == TrueLiteral:
+                return expression.right
+            if expression.right == TrueLiteral:
+                return expression.left
+            else:
+                return expression
+        else:
+            return expression
+        location_name, _ = context_field.location.get_location_name()
+        if location_name not in current_locations:
+            return TrueLiteral
+        else:
+            return expression
+    elif isinstance(expression, TernaryConditional):
+        if expression.predicate == TrueLiteral:
+            return expression.if_true
+        else:
+            return expression
+    elif isinstance(expression, BetweenClause):
+        lower_bound = expression.lower_bound
+        upper_bound = expression.upper_bound
+        if isinstance(lower_bound, ContextField) or isinstance(upper_bound, ContextField):
+            raise AssertionError(u'Found BetweenClause with ContextFields as lower/upper bounds. '
+                                 u'THis should never happen: {}'.format(expression))
+    else:
+        return expression
+
+
+def _construct_update_tags_visitor_fn(current_locations):
+    """Return an Expression updater using the given `current_locations`."""
+    def visitor_fn(expression):
+        return _update_tagged_expression(expression, current_locations)
+    return visitor_fn
+
+
+def _get_current_locations_from_match_traversals(match_traversals):
+    """Return a set of locations present in the given match traversals."""
+    current_locations = set()
+
+    for traversal in match_traversals:
+        for step in traversal:
+            if step.as_block is not None:
+                location_name, _ = step.as_block.location.get_location_name()
+                current_locations.add(location_name)
+    return current_locations
+
+
+def _lower_filters_in_match_traversals(match_traversals, visitor_fn):
+    """Return new match traversals, replacing filters involving non-existent tags with True."""
+    new_match_traversals = []
+    for traversal in match_traversals:
+        new_match_traversal = []
+        for step in traversal:
+            if step.where_block is not None:
+                new_filter = step.where_block.visit_and_update_expressions(visitor_fn)
+                if new_filter.predicate == TrueLiteral:
+                    new_filter = None
+                new_step = step._replace(where_block=new_filter)
+            else:
+                new_step = step
+            new_match_traversal.append(new_step)
+        new_match_traversals.append(new_match_traversal)
+    return new_match_traversals
+
+
+def lower_tagged_expressions_in_compound_match_query(compound_match_query):
+    """Lower Expressons involving non-existent ContextFields (tags)."""
```

Comment:
Try not to refer to tags in files that deal with IR, since the IR does not contain such a concept. This includes in function names, in comments, and in docstrings.